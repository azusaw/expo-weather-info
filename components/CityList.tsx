import React, { useState } from "react";
import { StyleSheet, TouchableOpacity } from "react-native";
import CityLocations from "@/assets/json/city-location.json";
import Colors from "@/constants/Colors";
import { Text, View } from "./Themed";
import { useLocationStore } from "@/store/useLocationStore";

const CityList = React.memo(({ onClose }: { onClose: () => void }) => {
  const { setLocation } = useLocationStore();
  const [selectedCityIndex, setSelectedCityIndex] = useState<number>();

  const onSelect = async (index: number) => {
    setSelectedCityIndex(index); // for showing active color
    await setLocation(CityLocations[index]);
    onClose();
  };

  return (
    <View style={styles.container}>
      <View style={styles.menuBox}>
        {CityLocations.map(({ name }, index) => (
          <TouchableOpacity key={name} onPress={() => onSelect(index)}>
            <View
              style={
                index === selectedCityIndex
                  ? styles.activeItem
                  : styles.menuItem
              }
            >
              <Text size={18}>{name}</Text>
            </View>
            {index < CityLocations.length - 1 && <View style={styles.border} />}
          </TouchableOpacity>
        ))}
      </View>
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    position: "relative",
    marginHorizontal: "auto",
    width: "100%",
    maxWidth: 400,
  },
  menuBox: {
    position: "absolute",
    bottom: 50,
    right: 5,
    width: 220,
    borderRadius: 5,
    backgroundColor: "#fff",
    overflow: "hidden",
  },
  menuItem: {
    padding: 15,
    borderRadius: 5,
  },
  activeItem: {
    padding: 15,
    backgroundColor: Colors.primary.default,
    borderLeftWidth: 6,
    borderLeftColor: Colors.secondary.default,
  },
  border: {
    height: 1,
    width: "100%",
    backgroundColor: Colors.primary.default,
  },
});

export default CityList;
