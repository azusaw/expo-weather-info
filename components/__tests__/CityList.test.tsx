import { fireEvent, render, waitFor } from "@testing-library/react-native";
import CityList from "@/components/CityList";
import CityLocations from "@/assets/json/city-location.json";

jest.mock("react-native/Libraries/Animated/Easing", () => ({
  bezier: jest.fn(),
}));

const mockProps = {
  onChange: jest.fn,
};

describe("CityList", () => {
  test("it renders CityList successfully", () => {
    const { toJSON } = render(<CityList {...mockProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("it renders all the city name", () => {
    CityLocations.forEach((city) => {
      const { getByText } = render(<CityList {...mockProps} />);
      expect(getByText(city.name)).toBeTruthy();
    });
  });

  test("it calls onChange when select item clicked", () => {
    const { getByText } = render(<CityList {...mockProps} />);
    const target = getByText(CityLocations[0].name);
    fireEvent.press(target);
    waitFor(() => expect(mockProps.onChange()).toHaveBeenCalled());
  });
});
