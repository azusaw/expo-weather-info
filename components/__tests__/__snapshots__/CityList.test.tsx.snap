// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityList it renders CityList successfully 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "inherit",
      },
      {
        "marginHorizontal": "auto",
        "maxWidth": 400,
        "position": "relative",
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "inherit",
        },
        {
          "backgroundColor": "#fff",
          "borderRadius": 5,
          "bottom": 50,
          "overflow": "hidden",
          "position": "absolute",
          "right": 5,
          "width": 220,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Tokyo
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Paris
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Hong Kong
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Bangkok
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Saint Petersburg
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Kiev
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Berlin
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Dublin
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          London
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          New York
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "backgroundColor": "#ddd",
              "height": 1,
              "width": "100%",
            },
          ]
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "inherit",
            },
            {
              "borderRadius": 5,
              "padding": 15,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#888",
                "fontFamily": "Montserrat_400Regular",
                "fontSize": 18,
              },
              undefined,
            ]
          }
        >
          Mexico City
        </Text>
      </View>
    </View>
  </View>
</View>
`;
